### ISSUE TITLE
ajax request, lockingmode=all ,still can't get session attribute sometimes

### ISSUE THREAD

**Issue by ghost** [magro/memcached-session-manager#257]
_From [zheng.gu...@gmail.com](https://code.google.com/u/105201542052703221351/) on November 11, 2014 07:50:57_

<b>What steps will reproduce the problem?</b>
 I just create random code for user to login, and put the random code into session attribute by using ajax,&#13;
but sometimes I can't find the attribute,&#13;
and sometimes I can get it.&#13;
&#13;
and I feel the problem may be related with "time".&#13;
if the random code was put into session for a while ,&#13;
I can find it every times. but if I don't wait and input the random code very quickly, the probability that the attribute can't be found may be very high 。&#13;
&#13;
I just used one tomcat server. &#13;
&#13;
I can't get any useful issues from now.&#13;
&#13;
&#13;
I used tomcat 7 .&#13;
in tomcat context.xml : &#13;
&#13;
  &lt;Manager className="de.javakaffee.web.msm.MemcachedBackupSessionManager"&#13;
 			memcachedNodes="n1:10.0.110.232:11211"&#13;
    			sticky="false"&#13;
			lockingMode="all"&#13;
			sessionBackupAsync="false"&#13;
			sessionBackupTimeout= "2"   		&#13;
			storageKeyPrefix="static:test"		&#13;
   			requestUriIgnorePattern=".*\.(ico|png|gif|jpg|css|js)$"&#13;
			transcoderFactoryClass="de.javakaffee.web.msm.serializer.kryo.KryoTranscoderFactory"&#13;
		&#13;
    		/&gt;  &#13;
&#13;
in lib,I used the following jars:&#13;
asm-3.2.jar&#13;
kryo-1.04.jar&#13;
kryo-serializers-0.11.jar&#13;
memcached-session-manager-1.8.3.jar&#13;
memcached-session-manager-tc7-1.8.3.jar&#13;
minlog-1.2.jar&#13;
msm-kryo-serializer-1.8.3.jar&#13;
reflectasm-1.01.jar&#13;
spymemcached-2.11.1.jar&#13;
&#13;

_Original issue: http://code.google.com/p/memcached-session-manager/issues/detail?id=218_

**Comment by google-code-exporter** [comment]
_From [zheng.gu...@gmail.com](https://code.google.com/u/105201542052703221351/) on November 11, 2014 00:01:47_

martin.grotzke：

 hi，I'm urgenly waiting for your help. 
and I tried <a title="multiple parallel requests for the same session with msm has problem" href="/p/memcached-session-manager/issues/detail?id=151">issue 151</a>, still can't work.&#13;

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on November 11, 2014 00:12:17_

Where do you read the value? Are you sure that this happens after the ajax request was completed?

**Comment by google-code-exporter** [comment]
_From [zheng.gu...@gmail.com](https://code.google.com/u/105201542052703221351/) on November 11, 2014 01:59:40_

I think I'm using msm in a very common context and should not meet these problems.

maybe any jar version is something wrong ?

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on November 11, 2014 14:01:54_

Can you activate debug logging according to the docs and share the log output?

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on November 12, 2014 00:15:39_

Do you have the related tomcat logs (with debug mode enabled)?

**Comment by google-code-exporter** [comment]
_From [zheng.gu...@gmail.com](https://code.google.com/u/105201542052703221351/) on November 20, 2014 08:47:42_

log file as attachment . 
 /wmt365/user!randomImage.action  create random and set to session attribute.
/wmt365/login!login.action check request random with session attribute random

I tested twice . 
the first is ok,the second is bad.
And I saw that ,the first time ,randomImage.action set attribute to the session,and when logined, 
"2014-11-21 00:36:44,257 [http-bio-80-exec-10] DEBUG de.javakaffee.web.msm.SessionIdFormat- Creating new session id with orig id 'ping' and memcached id 'n1'.
2014-11-21 00:36:44,260 [http-bio-80-exec-10] DEBUG de.javakaffee.web.msm.NodeAvailabilityCache- CacheLoader returned node availability 'true' for node 'n1'.
2014-11-21 00:36:44,261 [http-bio-80-exec-10] DEBUG de.javakaffee.web.msm.MemcachedBackupSessionManager$1- Loading session from memcached: F05F8770BB6E6CE2BB968C713088F840-n1"

Yes, session is loaded from memcached.

but the second time:
2014-11-21 00:36:50,449 [http-bio-80-exec-1] DEBUG de.javakaffee.web.msm.SessionIdFormat- Creating new session id with orig id 'ping' and memcached id 'n1'.
2014-11-21 00:36:50,452 [http-bio-80-exec-1] DEBUG de.javakaffee.web.msm.NodeAvailabilityCache- CacheLoader returned node availability 'true' for node 'n1'.
2014-11-21 00:36:50,460 [http-bio-80-exec-1] DEBUG de.javakaffee.web.msm.MemcachedBackupSessionManager$1- createSession invoked: F05F8770BB6E6CE2BB968C713088F840-n1
2014-11-21 00:36:50,460 [http-bio-80-exec-1] DEBUG de.javakaffee.web.msm.MemcachedBackupSessionManager$1- Created new session with id F05F8770BB6E6CE2BB968C713088F840-n1
2014-11-21 00:36:50,570 [http-bio-80-exec-1] DEBUG de.javakaffee.web.msm.BackupSessionService- Starting for session id F05F8770BB6E6CE2BB968C713088F840-n1

"Created new session" ,why created again ? not loaded from memached?


Can you help me out ? Thanks a lot.

**Attachment:** [catalina](http://code.google.com/p/memcached-session-manager/issues/detail?id=218#hc17)

**Comment by google-code-exporter** [comment]
_From [zheng.gu...@gmail.com](https://code.google.com/u/105201542052703221351/) on November 20, 2014 22:54:59_

martin.grotzke:

 Hello. I found the problem about the LRUCache.

====
  In MemcachedSessionService.java :
   private final LRUCache<String, Boolean> _invalidSessionsCache = new LRUCache<String, Boolean>( 2000, 5000 );
====


you defined the cache time to 5000,that's 5 seconds long.

When I logouted ,no sticky session called removed.
MemcachedBackupSessionManager- remove invoked, removeFromMemcached: true

thus ,the sessionid is input to the _invalidSessionsCache.

in short time(less than 5 seconds) after random created and input to session attribute,I relogined, but the session id was indicated to invalid one. 
then recreated the session ,and the random session attribute was also dispeared.

Thus, I have two ways :
1: reduce the LRUCache timeout from 5000 to 500 or more less ?
2: like another issue, remove the cookie ,every session with different cookie

**Comment by RemyaPR** [comment]
Is this issue resolved? We also face a problem that session attributes are lost sometimes.
